src/AppInstallerCLITests/Correlation.cpp

```diff
@@ -0,0 +1,323 @@
+{
+    Manifest manifest;
+    manifest.DefaultLocalization.Add<Localization::PackageName>(testCase.AppName);
+    manifest.DefaultLocalization.Add<Localization::Publisher>(testCase.AppPublisher);
+    manifest.Localizations.push_back(manifest.DefaultLocalization);
+    return manifest;
+}
+
+ARPEntry GetARPEntryFromTestCase(const TestCase& testCase)
+{
+    Manifest arpManifest;
+    arpManifest.DefaultLocalization.Add<Localization::PackageName>(testCase.ARPName);
+    arpManifest.DefaultLocalization.Add<Localization::Publisher>(testCase.ARPPublisher);
+    arpManifest.Localizations.push_back(arpManifest.DefaultLocalization);
+    return ARPEntry{ TestPackageVersion::Make(arpManifest), false };
+}
+
+void ReportMatch(std::string_view label, std::string_view appName, std::string_view appPublisher, std::string_view arpName, std::string_view arpPublisher)
+{
+    WARN(label << '\n' <<
+        "\tApp name      = " << appName << '\n' <<
+        "\tApp publisher = " << appPublisher << '\n' <<
+        "\tARP name      = " << arpName << '\n' <<
+        "\tARP publisher = " << arpPublisher);
+}
+
+ResultSummary EvaluateDataSetWithHeuristic(const DataSet& dataSet, const ARPCorrelationAlgorithm& correlationAlgorithm, bool reportErrors = false)
+{
+    ResultSummary result{};
+
+    // Each entry under test will be pushed at the end of this
+    // and removed at the end.
+    auto arpEntries = dataSet.ARPNoise;
+
+    for (const auto& testCase : dataSet.TestCases)
+    {
+        arpEntries.push_back(GetARPEntryFromTestCase(testCase));
+        auto match = correlationAlgorithm.GetBestMatchForManifest(GetManifestFromTestCase(testCase), arpEntries);
+        arpEntries.pop_back();
+
+        if (match)
+        {
+            auto matchManifest = match->Entry->GetManifest();
+            auto matchName = matchManifest.DefaultLocalization.Get<Localization::PackageName>();
+            auto matchPublisher = matchManifest.DefaultLocalization.Get <Localization::Publisher>();
+            if (matchName == testCase.ARPName && matchPublisher == testCase.ARPPublisher)
+            {
+                ++result.TrueMatches;
+            }
+            else
+            {
+                ++result.FalseMatches;
+
+                if (reportErrors)
+                {
+                    ReportMatch("False match", testCase.AppName, testCase.AppPublisher, matchName, matchPublisher);
+                }
+            }
+        }
+        else
+        {
+            if (testCase.IsMatch)
+            {
+                ++result.FalseMismatches;
+
+                if (reportErrors)
+                {
+                    ReportMatch("False mismatch", testCase.AppName, testCase.AppPublisher, testCase.ARPName, testCase.ARPPublisher);
+                }
+            }
+            else
+            {
+                ++result.TrueMismatches;
+            }
+        }
+    }
+
+    return result;
+}
+
+void ReportResults(ResultSummary results)

```

Comment:
I would include some details on the time that this is taking as well, specifically the average time to do a correlation.  Most useful in the "full" test case where there are multiple ARP entries to choose from, reporting the average time it took to check what would be a real world correlation would be good.

Reply:
Added total and average time to what is reported here.